### PR TITLE
test hygiene: track stray suites, cover shared utils, atomic env writes (PyTest - 4)

### PR DESCRIPTION
**Pull request by Manik-Khajuria-5**
## Description
Test-suite hygiene plus one durability fix in `JSONStore`.

- Moved two untracked suites into the collected tree: build verification  (`integration/test_build_verification.py`) and t-SNE (`unit/test_tsne.py`).
- Added `unit/test_shared_utils.py`; extended `unit/test_data_model.py` with `LazyEnvData`'s mapping and deep-copy behaviour.
- **Fix:** `serialize_env` wrote straight to the target file, so an interrupted save truncated the old environment and `load_env` reported it as empty. Writes now stage into `<path>.tmp` and `os.replace`; `save_undo` shares the same helper.

- Moved `visual_check.py` to `example/manual/` it needs a live server and a human eye, so it isn't collectable. Rule documented in `.agents/context/testing.md`.

## Integration note

This PR should be merged after #1671 , as it depends on the lower-layer changes introduced there.

## Motivation and Context

Two test files existed but were never collected, so that code was untested in CI. The atomic write closes a real data-loss window on env save.

## How Has This Been Tested?
`pytest py/tests` on Python 3.12: 593 passed (174 in the touched files). No new dependencies; all tests hermetic.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code refactor or cleanup (changes to existing code for improv)

## Checklist:
- [ ] I adapted the version number under `py/visdom/VERSION` according to [Semantic Versioning](https://semver.org/)
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

## Summary by Sourcery

Strengthen visdom’s persistence and update semantics while formalizing the Python test layout into unit and integration suites with shared test utilities and manual examples.

New Features:
- Add a configurable in‑memory cap for plot history panes, wired through application defaults and handler attributes.

Bug Fixes:
- Make JSONStore environment and undo saves atomic via staged temporary files and os.replace to avoid truncated or lost environments on interrupted writes.
- Fix plot update handling for categorical axes and all‑missing updates using a shared missing‑value helper, preventing TypeError and ensuring gaps are skipped safely.
- Ensure window indices are monotonically assigned based on the current maximum instead of len(env) so indices remain unique after deletions.
- Replace assertion‑based failures in fork_env, window read, and named trace updates with HTTP 400 errors carrying clear reasons, and handle empty update payloads without index errors.

Enhancements:
- Refactor update handlers to support capped plot history, robust trace injection when all traces have been deleted, and safer marker updates.
- Introduce shared test utilities (HTTP base case, payload builders, fakes and spy storage) and reorganize tests into unit and integration subpackages with common fixtures.
- Clarify pyproject pytest configuration with pythonpath support for testutils and registered markers for unit, integration, slow, and server tests.
- Exclude tests packages from the distributed visdom wheel to avoid shipping internal test code to end users.
- Document test layout, hermeticity rules, HTTP test patterns, and manual visual checks via updated testing context docs and example/manual assets.

Build:
- Adjust setup.py package discovery to exclude tests packages from distribution.
- Update pyproject.toml pytest configuration to add py/tests to pythonpath and register unit, integration, slow, and refined server markers.

Documentation:
- Expand testing documentation to describe the py/tests layout, fixtures, markers, HTTP test usage, and the role of manual visual checks.
- Add example/manual documentation and a visual_check script for human‑driven UI verification outside the automated pytest suite.

Tests:
- Add extensive unit and integration coverage for window construction, update semantics across pane types, environment and window lifecycles, edge cases, and shared utilities including tsne helpers and warning behavior.
- Introduce durability tests for JSONStore env and undo writes to validate atomic write behavior and long‑name hash fallback interactions.
- Add build verification integration tests to assert that static frontend assets are present and correctly served by the Application.

## Summary by Sourcery

Harden environment persistence and organize comprehensive hermetic coverage across unit, integration, and manual checks.

Bug Fixes:
- Make environment and undo persistence atomic so interrupted writes cannot truncate existing data.

Enhancements:
- Improve plot update handling for partial and all-missing trace data.
- Formalize the automated test layout with collectable unit and integration suites, shared utility coverage, and durability tests.

Documentation:
- Document hermetic pytest requirements and move human-driven visual checks into a dedicated manual examples area.

Tests:
- Add build verification, shared utility, t-SNE, lazy environment, deep-copy, and JSONStore durability coverage.

Chores:
- Move the visual verification script out of the pytest tree.

### CODE REVIEW

**Review by tonypzy — APPROVED** [review]

File: py/tests/integration/build_verification.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+TRACKED_ASSETS = [
+    "index.html",
+    "login.html",
+    "error.html",
+    os.path.join("js", "main.js"),
+    os.path.join("js", "main.js.map"),
+    os.path.join("css", "style.css"),
+    os.path.join("css", "login.css"),
+    os.path.join("css", "error.css"),
+    os.path.join("css", "network.css"),
```
Comment (rajnisht7) [inline py/tests/integration/build_verification.py:48]: there is no network.css in the repo, whereas two css files base.css and hparams.css are not tracked

File: py/tests/integration/build_verification.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    os.path.join("css", "rc-tree-select-overrides.css"),
+]
+
+# Written by build.py at install time; absent in a plain checkout.
+DOWNLOADED_ASSETS = [
+    "js/jquery.min.js",
+    "js/d3.v3.min.js",
+    "css/bootstrap.min.css",
```
Comment (rajnisht7) [inline py/tests/integration/build_verification.py]: kindly remove the assets which are not longer downloaded and update the tests accordingly

**Review by Jayantparashar10 — APPROVED** [review]

**Review by Manik-Khajuria-5 — COMMENTED** [review]
@rajnisht7 Can u please review this PR again I have made changes acc to ur suggestions

File: py/tests/integration/build_verification.py (hunk 1 above)
Comment (Manik-Khajuria-5) [reply py/tests/integration/build_verification.py:48]: done

File: py/tests/integration/build_verification.py (hunk 2 above)
Comment (Manik-Khajuria-5) [reply py/tests/integration/build_verification.py]: done

**Review by rajnisht7 — APPROVED** [review]

**Review by Saksham-Sirohi — APPROVED** [review]
